### PR TITLE
Add document for checklist when adding new components

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity documentation
/kind enhancement

**What this PR does / why we need it**:
This PR adds a new document for developers with a checklist for what to pay attention to when adding new components to garden, seed, or shoot clusters.

**Release note**:

```feature developer
A new document for developers has been added with a checklist for what to pay attention to when adding new components to garden, seed, or shoot clusters. Read it [here](https://github.com/gardener/gardener/blob/master/docs/development/component-checklist.md).
```

### CODE REVIEW

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md:27]: I think it should be mentioned that some configs must not be unique, for example monitoring configuration. The reason is that the old revision stays in the cluster and it is still aggregated in the prometheus config.

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md:39]: Some extension controllers are also admission webhooks, can you mention how are their certificates managed when they are installed via ControllerDeployment?

**Review by vpnachev — COMMENTED** [review]
Very nice!

File: docs/development/component-checklist.md
Comment (rfranzke) [reply docs/development/component-checklist.md:39]: OK, just to clarify: You don't mean the admission binaries running in the garden cluster for validation (because here certificates are managed by human operators during deployment), but the regular extensions running in the seed which serve webhook endpoints (here, the extensions library manages this out-of-the-box)?

Comment (vpnachev) [reply docs/development/component-checklist.md:39]: Yes, I mean the the webhooks running in the seed clusters, e.g. the one running this functions https://github.com/gardener/gardener-extension-provider-aws/blob/47d6bb34a538f3dfeedcf99361696de72d1eeae2/pkg/webhook/controlplane/ensurer.go#L74

File: docs/development/component-checklist.md
Comment (timebertt) [inline docs/development/component-checklist.md:13]: You could mention the use of the go embed feature. 
E.g., when helm charts are used like for istio, they should be embedded into the binary instead of filesystem-based.
Also, generated templates (e.g., CRDs) are supposed to be embedded.

File: docs/development/component-checklist.md
Comment (timebertt) [inline docs/development/component-checklist.md]: You could mention the benefits that come with using `ManagedResources`, e.g., support for "delete on invalid updates", health checks out of the box, etc.

File: docs/development/component-checklist.md
Comment (timebertt) [inline docs/development/component-checklist.md]: You could reference https://github.com/gardener/gardener/blob/d22a2154613fb58318a310b96a4bc524c5e17bd9/docs/deployment/image_vector.md (which explains the overwrite and other advanced image vector topics)

File: docs/development/component-checklist.md
Comment (timebertt) [inline docs/development/component-checklist.md]: You could add an explanation, that unique configmaps/secrets are superior to checksum annotations on the pod template and we try to use unique objects instead.

**Review by timebertt — COMMENTED** [review]
This document is awesome, thank you!
A few comments from a first glance of the first section.
Will continue with the other sections.

File: docs/development/component-checklist.md
Comment (timebertt) [inline docs/development/component-checklist.md:62]: Nit: when reading "shoot access token" devs might think, they can only use it for shoot clusters. But this will also be possible for components running in the garden runtime cluster talking to the virtual garden cluster.

**Review by timebertt — COMMENTED** [review]
Again, awesome document!
Not much to add :)

File: docs/development/component-checklist.md
Comment (oliver-goetz) [inline docs/development/component-checklist.md]: It would be helpful having names for the proper deployment ways. There seems to be `gardener-resource-manager` and two other options. The later options are not very tangible.
Additionally, it would be helpful to map the examples to the deployment options.

**Review by oliver-goetz — COMMENTED** [review]
The `General` section could also include an example of how to structure the data types of a component (e.g. there should be a `Values` structure similar to what is written into a helm values.yaml) - [ref](https://github.com/gardener/gardener/blob/6a0fea86850ffec8937d1956bdf1a8ca6d074f3b/pkg/operation/botanist/component/metricsserver/metrics_server.go#L80-L97) 

Additionally we could mention how a new component should be structured (e.g. create a separate `logging.go` file for filters etc.) and provide a good example for it.

File: docs/development/component-checklist.md
Comment (rfranzke) [reply docs/development/component-checklist.md]: I'll include a teaser but refer to the controller documentation since this is the better place to talk about the features and benefits.

File: docs/development/component-checklist.md
Comment (rfranzke) [reply docs/development/component-checklist.md:62]: Well, for now, this is the status quo. I would rather keep it for now to not add information here which has not become reality yet. We can adapt this document when this feature (GRM for virtual garden) is supported by `gardener-operator`.

File: docs/development/component-checklist.md
Comment (rfranzke) [reply docs/development/component-checklist.md]: Not sure whether I can come up with some artificial names right now 😅 Any suggestions? The text in this section should explain the three cases (direct application with client, GRM only, or mixed scenario with the utility functions). Isn't this good enough/what do you miss?

File: docs/development/component-checklist.md
Comment (rfranzke) [reply docs/development/component-checklist.md:39]: I tried to incorporate this, but I reverted it again since it doesn't seem to be the right place to talk about this. This document is about a checklist for things to keep in mind when a component deployed via Gardener (or an extension). Talking now about Gardener/extension-specifics is probably confusing and doesn't belong here.

File: docs/development/component-checklist.md
Comment (timebertt) [reply docs/development/component-checklist.md:62]: Agreed

**Review by timebertt — APPROVED** [review]
Thanks!
/lgtm

/hold
for the other reviews

File: docs/development/component-checklist.md
Comment (acumino) [inline docs/development/component-checklist.md]:
```suggestion
   Hence, the image references must not be hard-coded in the pod template spec but read from this so-called [image vector](https://github.com/gardener/gardener/blob/6a0fea86850ffec8937d1956bdf1a8ca6d074f3b/docs/deployment/image_vector.md) instead.
```

**Review by acumino — COMMENTED** [review]
One nit.
/lgtm

**Comment by rfranzke** [comment]
Thanks for all the reviews. Is there anybody else who would like to take a final look before we merge this?

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md:132]: Can you add link to the monitoring documentation, I think it is https://github.com/gardener/gardener/blob/master/docs/extensions/logging-and-monitoring.md#extensions-monitoring-integration.

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md:138]: Can you add link to the logging documentation, I think it is https://github.com/gardener/gardener/blob/master/docs/extensions/logging-and-monitoring.md#fluent-bit-log-parsers-and-filters.

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md]:
```suggestion
   All other resources (seed or shoot system components) are deployed via `gardener-resource-manager`'s [Resource controller](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#managedresource-controller) (`ManagedResource`s) since it performs health checks out-of-the-box and has a lot of other features (see its documentation for more information).
```

File: docs/development/component-checklist.md
Comment (vpnachev) [reply docs/development/component-checklist.md]:
```suggestion
   Hence, the image references must not be hard-coded in the pod template spec but read from this so-called ["image vector"](https://github.com/gardener/gardener/blob/master/docs/deployment/image_vector.md) instead.
```

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md]:
```suggestion
   Stale/unused `ConfigMap`s/`Secret`s are garbage-collected by `gardener-resource-manager`'s [GarbageCollector](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#garbage-collector-for-immutable-configmapssecrets).
```

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md]:
```suggestion
   This makes `gardener-resource-manager`'s [TokenInvalidator](https://github.com/gardener/gardener/blob/6a0fea86850ffec8937d1956bdf1a8ca6d074f3b/docs/concepts/resource-manager.md#tokeninvalidator) invalidating the static token secret and its [`ProjectedTokenMount` webhook](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#auto-mounting-projected-serviceaccount-tokens) injecting a projected token automatically.
```

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md]:
```suggestion
   Components which need to talk to a target cluster different from their runtime cluster (e.g., running in seed cluster but talking to shoot) then the `gardener-resource-manager`'s [TokenRequestor](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#tokenrequestor) should be used to manage a so-called "shoot access token".
```

File: docs/development/component-checklist.md
Comment (vpnachev) [inline docs/development/component-checklist.md]:
```suggestion
   To support high-availability deployments, `gardener-resource-manager`s [HighAvailabilityConfig](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#high-availability-config) webhook injects the proper specification like replica or topology spread constraints.
```

**Review by vpnachev — COMMENTED** [review]
I think it will be better links to documentation (.md files) to not be pinned to particular revision but to to the master branch. You can also try to use relative links, this way also the branch name will not be pinned.

**Comment by rfranzke** [comment]
> I think it will be better links to documentation (.md files) to not be pinned to particular revision but to to the master branch. You can also try to use relative links, this way also the branch name will not be pinned.

Nah, this way the links will get outdated very soon when lines change. We also follow this approach (pinning examples to specific version) in https://github.com/gardener/gardener/blob/master/docs/development/testing.md. This ensures the linked examples always fit to the explanation.

File: docs/development/component-checklist.md
Comment (oliver-goetz) [reply docs/development/component-checklist.md]: Naming the examples accordingly would already clarify it.
`direct application with client`, `gardener-resource-manager` and `mixed scenario` instead of `example [1|2|3]`

Comment (rfranzke) [reply docs/development/component-checklist.md]: done w/ 8fc37e64f

**Pull request merged**
